File: backend/app/models/schemas.py
from pydantic import BaseModel, EmailStr
from datetime import datetime
from enum import Enum
from typing import Optional

class VehicleType(str, Enum):
    CAR = "car"
    MOTORCYCLE = "motorcycle"
    BOAT = "boat"

class ProductBase(BaseModel):
    name: str
    description: Optional[str] = None
    price: float
    vehicle_type: VehicleType
    brand: str
    model: str
    year: int
    mileage: float
    is_for_sale: bool = True
    is_for_rent: bool = False

class ProductCreate(ProductBase):
    pass

class Product(ProductBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True  # Permet la conversion des modèles SQLAlchemy en schémas Pydantic

class RentalDetailsBase(BaseModel):
    monthly_price: float
    minimum_duration_months: int
    includes_insurance: bool = True
    includes_assistance: bool = True
    includes_maintenance: bool = True
    includes_technical_check: bool = True

class DossierBase(BaseModel):
    product_id: int
    dossier_type: str  # 'purchase' ou 'rental'
    down_payment: Optional[float] = None
    loan_amount: Optional[float] = None
    rental_duration: Optional[int] = None
    start_date: Optional[datetime] = None

class DossierCreate(DossierBase):
    pass

class Dossier(DossierBase):
    id: int
    user_id: int
    status: str
    submitted_at: datetime
    reviewed_at: Optional[datetime] = None
    reviewed_by: Optional[int] = None

    class Config:
        orm_mode = True

File: frontend/src/forms/schemas.ts
import { z } from "zod";

export const VehicleTypeSchema = z.enum(["car", "motorcycle", "boat"]);
export type VehicleType = z.infer<typeof VehicleTypeSchema>;

export const RentalDetailsSchema = z.object({
  monthly_price: z.number().min(0, "Le prix mensuel doit être positif"),
  minimum_duration_months: z.number().min(1, "Durée minimale de 1 mois"),
  includes_insurance: z.boolean().default(true),
  includes_assistance: z.boolean().default(true),
  includes_maintenance: z.boolean().default(true),
  includes_technical_check: z.boolean().default(true),
});

export const ProductSchema = z.object({
  name: z.string().min(1, "Le nom est obligatoire"),
  description: z.string().optional(),
  price: z.number().min(0, "Le prix doit être positif"),
  vehicle_type: VehicleTypeSchema,
  brand: z.string().min(1, "La marque est obligatoire"),
  model: z.string().min(1, "Le modèle est obligatoire"),
  year: z.number()
    .min(1900, "Année invalide")
    .max(new Date().getFullYear() + 1),
  mileage: z.number().min(0, "Le kilométrage doit être positif"),
  is_for_sale: z.boolean().default(true),
  is_for_rent: z.boolean().default(false),
  rental_details: RentalDetailsSchema.optional(),
  dossier_type: z.string().optional(), 
}).superRefine((data, ctx) => {
  if (data.is_for_rent && !data.rental_details) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "Les détails de location sont requis",
      path: ["rental_details"],
    });
  }
});
export type VehicleFormValues = z.infer<typeof ProductSchema>;

export const CategorySchema = z.object({
  name: z.string().min(1, "Le nom est obligatoire"),
  description: z.string().optional(),
});

export type CategoryFormValues = z.infer<typeof CategorySchema>;